lib/user_interface.rb

```diff
@@ -26,18 +46,37 @@ def run
 
     loop do
       user_input = input.gets.chomp
-      break if valid?(user_input)
+      break if valid_choise?(user_input)
 
       output.print ERROR_MESSAGE
     end
     user_input.to_i
   end
 
+  def ask_for_fields
+    contact_details = {}
+    FIELDS_TO_PROMPTS.each do |field, prompt|
```

Comment:
You can simplify this with [`each_with_object`](https://ruby-doc.org/core-2.7.1/Enumerable.html#method-i-each_with_object), which allows you to directly create the hash and return it on the fly without having to create `contact_details` as an outer variable:

```ruby
def ask_for_fields
  FIELDS_TO_PROMPTS.each_with_object({}) do |(field, prompt), contact_details|
    # do something with contact_details
  end    
end
```

Later on:

```ruby
contact_details = ui.ask_for_fields
=> {
  name: 'Foo',
  ...
}
```

Reply:
![image](https://user-images.githubusercontent.com/60718290/93900084-e7d1f700-fcec-11ea-874d-155545e0254a.png)
this is so hard to understand

Reply:
* memo object is `contact_details`, is the object you keep in memory while you add stuff to it in every loop

* the initially given object is the one you pass as an argument of `each_with_object`, which is going to be the initial value of the memo object. In this case, we start with an empty hash, because `contact_details` is going to be a hash, and before we start iterating over it, it is empty.

* `returns anything assigned to this memo object` is not exactly what is going on, the thing you pass as an argument (the empty hash in this case) is what is returned, the memo object is just a temporary object where the thing you pass as an argument is copied to while you are iterating.

* There are no unused arguments in this construction

In the example they give you, they are iterating on an array rather than a hash. The array to be returned (`evens`) is passed as an argument to `each_with_object`, as you can see they start with an empty array. Then this gets copied to the memo object `a`, which is the temporary array created in memory to use while we iterate on it. When the iterations are finished, the thing you passed as an argument is returned (the empty array, which now is not empty anymore) and saved in `evens`.

You don't always have to pass an empty array or an empty hash to `each_with_object`, sometimes it is already initialized with some value.